src/client.c

```diff
@@ -161,6 +161,12 @@ static void start_auth_query(PgSocket *client, const char *username)
 	if (!auth_db)
 		return;
 
+    if (auth_db->admin) {
```

Comment:
Additionally, I wonder if we should also prevent users from setting auth_dbname as `pgbouncer` in the query.

The check we perform on [parsing database](https://github.com/pgbouncer/pgbouncer/blob/master/src/loader.c#L216) name can also apply to [auth_dbname](https://github.com/pgbouncer/pgbouncer/blob/master/src/loader.c#L144). This could avoid explicit else block.

Reply:
> Could be nicer to have this in `prepare_auth_database`. For instance:
> 
> ```
>   if (!auth_dbname)
> 	auth_db = client->db;
>   else
> 	auth_db = find_database(auth_dbname);
> 
> <other checks>
> 
>    if (auth_db->admin) {
>         slog_error(client, "admin database \"%s\" cannot be an authentication database", auth_db->dbname);
>         disconnect_client(client, true, "bouncer config error");
>         return null;
>    }
> ```

That is a reasonable point, I agree that these checks should be within the `prepare_auth_database` function. I will redo it.

Reply:
> Additionally, I wonder if we should also prevent users from setting auth_dbname as `pgbouncer` in the query.
> 
> The check we perform on [parsing database](https://github.com/pgbouncer/pgbouncer/blob/master/src/loader.c#L216) name can also apply to [auth_dbname](https://github.com/pgbouncer/pgbouncer/blob/master/src/loader.c#L144). This could avoid explicit else block.


If I understand you clearly, you meant that if we set this check in the set_auth_dbname we will avoid explicit else in the prepare_auth_database. In my point of view, it will throw the same error, when we use fallback databases. For example, we can write in the config `* = host=127.0.0.1 port=6432` and then try to connect to pgbouncer DB using `psql -h localhost -p 6432 -U stats -d pgbouncer`, it will force the pgbouncer exiting with the same error. I suggest keeping this check for both cases (with explicit else) even if we will fix find another solution to prevent using pgbouncer as auth_dbname.

Because of that, I added an additional test when we set up a fallback DB connection with an admin database (pgbouncer DB).

Reply:
We still have two inconvenient behaviors:
* No cf_auth_dbname checks
* No checks for fallback DB

I didn't find a convenient solution for parsing [pgbouncer] section which I could implement with the current parser and it wouldn't be a shame. I don't want to mix parsing logic with kind of business logic. Anyway, I should give a few ideas:

1. We could define new type macros in [cfparser](https://github.com/libusual/libusual/blob/b0bc727b793b6a94902b46d06ad87e70eb2a2e0c/usual/cfparser.h#L178) which describes `auth_dbname` constraints. It would be `#CF_AUTHDB`:
```
#define CF_STR	{ cf_set_authdb, cf_get_authdb }
```
And `cf_set_authdb` ([placed here](https://github.com/libusual/libusual/blob/b0bc727b793b6a94902b46d06ad87e70eb2a2e0c/usual/cfparser.c#L441)) would be a wrapper that gives us additional checks, such as:
```
bool cf_set_str(struct CfValue *cv, const char *value)
{
	if (strcmp(dbname, "pgbouncer") != 0) {
		log_error("cannot use the reserved \"%s\" database as an auth_dbname", new_auth_dbname);
		return false;
	}
	return cf_set_str(cv, value);
}
```
I clearly understand that this is not the way to handle it, because it is dependency.

2. Another option is to add an additional check after file parsing. For example, add a check function
```
bool loader_auth_dbname_check(void)
{
    if (cf_auth_dbname && strcmp(cf_auth_dbname, "pgbouncer") == 0) {
        log_error("cannot use the reserved \"%s\" database as an auth_dbname", cf_auth_dbname);
        free(cf_auth_dbname);
        cf_auth_dbname = NULL;
        return false;
    }
    return true;
}
```
And place additional check [before this line](https://github.com/pgbouncer/pgbouncer/blob/master/src/main.c#L409):
```
if (ok){
   ok = loader_auth_dbname_check();
}
```
It looks weak too.

Reply:
In my point of view, we have two things that should be implemented somehow:
* configuration parameters constraints
* config reloading with shadow copy approach

Additionally, I suppose we should implement a config testing utility, like `nginx -t`. Because now if we add such constraints (or just made a simple mistake in INI) we may partially override the old settings which may bring unpredictable behavior. I suppose we can implement something in the next PR's but It will bring us huge refactoring.

Reply:
Added validations that cover:
* config:
  * [pgbouncer] section
  * [databases] section
    * validation for all explicitly defined DB's
    * validation for authdb connection str
* runtime:
  * Check inside the `prepare_auth_database` function